cranelift-codegen/src/isa/x86/enc_tables.rs

```diff
@@ -99,6 +125,136 @@ fn size_plus_maybe_sib_or_offset_for_in_reg_1(
     sizing.base_size + additional_size_if(1, inst, divert, func, needs_sib_byte_or_offset)
 }
 
+/// Infers whether a dynamic REX prefix will be emitted, for use with one in-reg.
+///
+/// A REX prefix is known to be emitted if either:
+///  1. The EncodingBits specify that REX.W is to be set.
+///  2. Registers are used that require REX.R or REX.B bits for encoding.
+fn size_with_inferred_rex_for_one_in_reg(
+    sizing: &RecipeSizing,
+    enc: Encoding,
+    inst: Inst,
+    divert: &RegDiversions,
+    func: &Function,
+) -> u8 {
+    // Emit a REX prefix if REX.W is requested by the encoding.
+    let bits = EncodingBits::from(enc.bits());
+    if u8::from(bits.rex_w()) > 0 {
+        return sizing.base_size + 1;
+    }
+    sizing.base_size + additional_size_if(0, inst, divert, func, needs_rex_rb)
+}
+
+/// Infers whether a dynamic REX prefix will be emitted, for use with two in-regs.
+///
+/// A REX prefix is known to be emitted if either:
+///  1. The EncodingBits specify that REX.W is to be set.
+///  2. Registers are used that require REX.R or REX.B bits for encoding.
+fn size_with_inferred_rex_for_two_in_regs(
+    sizing: &RecipeSizing,
+    enc: Encoding,
+    inst: Inst,
+    divert: &RegDiversions,
+    func: &Function,
+) -> u8 {
+    // Emit a REX prefix if REX.W is requested by the encoding.
+    let bits = EncodingBits::from(enc.bits());
+    if u8::from(bits.rex_w()) > 0 {
+        return sizing.base_size + 1;
+    }
+
+    // Emit a REX prefix if REX.B is needed to encode the first register.
+    let reg0 = divert.reg(func.dfg.inst_args(inst)[0], &func.locations);
+    if needs_rex_rb(reg0) {
+        return sizing.base_size + 1;
```

Comment:
This applies to all the `size_with_inferred_rex*` here - how do you feel about structuring bodies like:
```
let needs_rex = (bits.rex_w() != 0) || // it'd be handy if `rex_w` could be obtained as a bool, alas
  test_input(0, inst, divert, func, needs_rex_rb) ||
  test_input(1, inst, divert, func, needs_rex_rb);

sizing.base_size + if needs_rex { 1 } else { 0 }
```
? Where `input_test` and `output_test` are `additional_size_if` but returning only the predicate? Admittedly I'm not a huge fan of `additional_size_if` because it's a bit eager in turning "we need a prefix" into "it will be so-and-so many bytes" (which would be complicating with @abrown's mention of VEX/EVEX!), so having a generic "is this true for this operand" which we can use to later decide "we do want a REX prefix, and that's one byte", seems nice. I also personally like avoiding the early returns with hardcoded sizes, YMMV!

Reply:
I like it. Will make that change.